### ISSUE TITLE
One key sequence doesn't work, others do

### ISSUE THREAD

**Issue by Anonymouse83** [autokey/autokey#821]
### AutoKey is a Xorg application and will not function in a Wayland session. Do you use Xorg (X11) or Wayland?

Unknown

### Has this issue already been reported?

- [X] I have searched through the existing issues.

### Is this a question rather than an issue?

- [X] This is not a question.

### What type of issue is this?

Bug

### Choose one or more terms that describe this issue:

- [ ] autokey triggers
- [X] autokey-gtk
- [ ] autokey-qt
- [ ] beta
- [X] bug
- [ ] critical
- [ ] development
- [ ] documentation
- [ ] enhancement
- [ ] installation/configuration
- [X] phrase expansion
- [X] scripting
- [ ] technical debt
- [ ] user interface

### Which Linux distribution did you use?

Operating System: Ubuntu 22.04.2 LTS              
          Kernel: Linux 5.19.0-35-generic
    Architecture: x86-64

### Which AutoKey GUI did you use?

GTK

### Which AutoKey version did you use?

0.95.10

### How did you install AutoKey?

sudo apt install

### Can you briefly describe the issue?

I wanted to provide a shortcut to access Thunderbird's running Mail filters on a folder - which requires the key combination of <alt>t and then "r".
I have several phrases installed and working, so I know the principle works, so I thought that a couple of seconds later, it would be fine! Oh dear! Silly me! And I fully accept this is most likely to be me being a numpty!
I have tried `<alt>+t+r`, `<alt>t+r`, `<alt>tr`, ... but none work. (However, this sequence works as detailed above!)
So, I thought I would see what happens if I ran the shortcut in a text editor, and I would get e.g. "tr" as the inserted values - which explained why it wasn't working in TB.
So, I pondered if this was due to needing to release the <alt>t keys before pressing the "r" key - and so I wrote a script (well, I wrote numerous variations, some mimicking the `<alt>tr` combinations, as well as releasing the `<alt>t` keys before pressing the "r" key, and equally, none worked.
Examples include:-
`# keyboard.send_key("<alt>+t")
# keyboard.send_key("r")
keyboard.press_key("<alt>")
keyboard.send_key("t")
keyboard.release_key("<alt>")
keyboard.send_key("r")`

I even tried to create the script from the keyboard/ mouse strokes, but that failed as well.

I am completely at a loss as to how to get this to work, so any ideas would be most appreciated.

Many thanks!

### Can the issue be reproduced?

Always

### What are the steps to reproduce the issue?

Create the phrase `<alt>t+r` or the equivalent script as outlined above, and try to run the script/ phrase from the main screen in TB - which does require one to be in a folder, and to have filters that will have something to do (or else how does one know it has worked?!).
If the folder is empty, the menu called up by `<alt>t` will show the run filter on folder option ("r") to be greyed out.
In most cases, I found that TB would ask if I wanted to move to the next unread message.

### What should have happened?

I expected the filters to be applied.

### What actually happened?

Filters were not applied, but TB would usually prompt to ask if I wanted to move to the next unread message.

### Can you provide the output of the AutoKey command?

```bash
If I run it in a text editor to see the output, usually, I see "tr" or with the <alt>t+r I see "+r".

Debug for <alt>t+r shown below:

2023-03-25 09:30:26,702 DEBUG - iomediator - Key.ALT pressed
2023-03-25 09:30:27,969 DEBUG - service - Raw key: 'j', modifiers: [<Key.ALT: '<alt>'>], Key: j
2023-03-25 09:30:27,969 DEBUG - service - Window visible title: '*Untitled Document 1 - gedit', Window class: 'gedit.Gedit'
2023-03-25 09:30:27,969 INFO - service - Matched Phrase "Remove Junk Mail" with hotkey and prompt=False
2023-03-25 09:30:27,970 DEBUG - iomediator - Send via event interface
2023-03-25 09:30:27,970 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/autokey/service.py", line 207, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
2023-03-25 09:30:27,973 DEBUG - interface - Send modified key: modifiers: ['<alt>'] key: t
2023-03-25 09:30:27,975 DEBUG - interface - Sending string: '+r'
2023-03-25 09:30:28,461 DEBUG - iomediator - Key.ALT released

<alt>+tr brings up the TB Tools menu, but nothing else - and in Gedit, leaves "r".
```

**Comment by josephj11** [comment]
Welcome to the AutoKey community, Anonymouse83.

It's primarily AutoKey that's being "numpty". :)

Unpacking this:

1. It's quite annoying that the Thunderbird add-ons to do this don't work any more. Once we get this working, I will add it to my system.

2. On my system, the required keystrokes are `<alt>+tR` With a capital R.

3. Take a look at [this](https://github.com/autokey/autokey/wiki/Emitting-Keyboard-Events) and [this](https://github.com/autokey/autokey/wiki/Key-Combinations) for how to do this correctly.

TL;DR: Don't use `keyboard.send_key` for anything unless you need its repeat parameter and don't mix press and release API calls with other keyboard API calls. It won't work.

Although it's not explicitly mentioned in the provided links, you can't mix asynchronous and synchronous API calls. This really isn't well documented. The send_key API calls and clipboard  calls are asynchronous. They start another thread and return succesfully immediately even though the work is not completed. Usually, adding some delays around the asynchronous calls gives them time to complete and work as expected, but when you mix them with synchronous API calls like press and release_key, they appear to operate in a separate context, so pressed keys have no effect once a subsequent asynchronous API call has been issued.

Handling contingencies such as disabled application options:

This is tricky because almost no applications are designed to be automated. If you want to go down that rabbit hole,  you can tell AutoKey to "look" at the screen and decide what to do based on what it finds. You can use the `visgrep()` API call for that, but may need to call Xautomatiion's full version of that command externally in some situations. That's a separate and longer discussion.

It won't affect this issue, but you should consider upgrading to AutoKey 0.96.0. It has lots of bug fixes and a few new features. [Source](https://github.com/autokey/autokey/releases/) - [Instructions](https://github.com/autokey/autokey/wiki/Installing). Since you're using a Debian derivative distro, Ubuntu, you can do this the easy way using our provided debs.

**Comment by Anonymouse83** [comment]
Thank you very much for your help.

I updated to AutoKey 0.96.0 as you recommended, and I changed the phrase to `<alt>+tR` (I haven't yet read the links you recommended - but will - thank you - and I am puzzled as to why it should be "R" rather than "r" - as "r" works "manually"!)

However, even with the changes, there is no difference in the results - i.e. running the phrase calls up the tools menu then stops!

Thank you so much for all the additional help - it is really appreciated.

The only reason I set up the script was in case that allowed me to generate the solution, but it was always my preference to use the phrase solution!

**Comment by josephj11** [comment]
Just going by what the menu shows me. Maybe it isn't case sensitive.

![Thunderbird+Tools](https://user-images.githubusercontent.com/4546046/227741396-0ae055d5-c79b-43ff-8e93-304709d0469e.png)

**Comment by josephj11** [comment]
The following script works on my system.

```
keyboard.send_keys("<alt>+t")
time.sleep(0.1)
keyboard.send_keys("R")
```

I set a window filter of ".*thunderbird.*" for safety and a hotkey of alt+j.

In Thunderbird, I created a new folder and copied a few messages into it. This gave me somewhere safe to test.

Next, I defined a Thunderbird message filter to be run manually only (again  for safety). It looks for messages tagged Important (1) and tags them as Work (2). I don't have any messages with both of those tags set, so if the filter gets loose, it can't do much damage.

I tagged one of the messages in the test folder as important.
Pressing Alt+j with either the folder selected or a message in the folder selected caused the Work (2) tag to be added to the message already tagged as Important (1), so the whole thing works.

I tried this as a phrase and it did not work. The most likely reason for this is that it takes Thunderbird some time after receiving alt+t to display the dropdown menu and prepare for further input. That's why I added the delay after that before sending the R. 0.1 seconds is my usual starting point for delays and since it worked, I didn't change it.

Thunderbird message filters are tricky. i have a large number of them. In the old days, there was the filtaquilla add-on that added things like being able to restrict a filter to only match messages in a particular folder, but they "improved" Thunderbird until they broke it and almost all the other extensions.

They won't admit it, but I firmly believe that Mozilla's slogan is "If it ain't broke, fix it until it is."

Don't even get me started on Firefox! I was in love with it up to version 3.x. After that, they started breaking things until version 60's scorched earth policy which killed all the must have extensions I relied upon. Now, I use Vivaldi and Brave for everything and only use Firefox when the other browsers won't work. Even then, I sometimes use Waterfox Classic.

**Comment by Anonymouse83** [comment]
I was going by what I type to execute it - and lowercase works manually - but still no joy with Autokey. I appreciate that I am using Autokey because TB doesn't seem to full execute the filters - or else I wouldn't need this phrase - but now it is a puzzle as to why it doesn't work - and so now will also raise a bug on TB!
Any other clues as to solving this would be most appreciated!
Many thanks!

**Comment by Anonymouse83** [comment]
Brilliant - thank you so much - that works!
I think I will also raise a bug on TB - as the filters should run, but only seem to partially run, and I have no idea why - and it only seems to impact the filters that are dealing with SPAM!
I will also look into the browsers you recommend - I don't trust Chrome, so have been loyal to Firefox for many years, but have been frustrated at the extensions that keep being removed!
My sincere thanks again,
